File: services/trivia_data.py
"""
Cached data loading from dotabase for trivia questions.
"""

from __future__ import annotations

import json
import re
from collections import Counter
from dataclasses import dataclass
from functools import lru_cache

from dotabase import Ability, Hero, Item, Response, dotabase_session
from sqlalchemy.orm import joinedload

_STEAM_CDN = "https://cdn.cloudflare.steamstatic.com/apps/dota2/images/dota_react"


# ---------------------------------------------------------------------------
# Dataclasses
# ---------------------------------------------------------------------------

@dataclass(frozen=True, slots=True)
class HeroData:
    id: int
    name: str  # internal name (npc_dota_hero_...)
    localized_name: str
    real_name: str | None
    hype: str | None
    bio: str | None
    attr_primary: str | None
    is_melee: bool
    base_movement: int | None
    base_armor: int | None
    attack_rate: float | None
    attr_str_gain: float | None
    attr_agi_gain: float | None
    attr_int_gain: float | None
    image_url: str | None
    armor_at_level1: float | None  # base_armor + agility_base / 6 (stat panel value at level 1)
    vision_night: int | None
    turn_rate: float | None
    attack_damage_min: int | None
    attack_damage_max: int | None


@dataclass(frozen=True, slots=True)
class AbilityData:
    id: int
    name: str
    localized_name: str
    hero_id: int | None
    hero_name: str | None
    damage_type: str | None
    damage: str | None
    cooldown: str | None
    lore: str | None
    scepter_upgrades: bool
    scepter_description: str | None
    shard_upgrades: bool
    shard_description: str | None
    innate: bool
    icon_url: str | None
    mana_cost: str | None


@dataclass(frozen=True, slots=True)
class ItemData:
    id: int
    localized_name: str
    cost: int | None
    lore: str | None
    neutral_tier: int | None
    icon_url: str | None
    is_neutral_enhancement: bool
    ability_special: str | None  # JSON string of bonus descriptions
    active_cooldown: str | None  # cooldown for active use (e.g. BKB, Blink)


@dataclass(frozen=True, slots=True)
class VoicelineData:
    hero_id: int
    text: str


@dataclass(frozen=True, slots=True)
class FacetData:
    id: int
    localized_name: str
    hero_id: int
    hero_name: str | None
    description: str | None


# ---------------------------------------------------------------------------
# CDN URL helpers
# ---------------------------------------------------------------------------

def hero_image_url(hero_name: str) -> str | None:
    """Build Steam CDN hero portrait URL from internal name."""
    slug = hero_name.replace("npc_dota_hero_", "")
    if not slug:
        return None
    return f"{_STEAM_CDN}/heroes/{slug}.png"


def ability_icon_url(icon_path: str | None) -> str | None:
    """Build Steam CDN ability icon URL from dotabase icon path."""
    if not icon_path:
        return None
    # e.g. /panorama/images/spellicons/antimage_mana_break_png.png
    slug = icon_path.replace("/panorama/images/spellicons/", "").replace("_png.png", "")
    if slug.endswith(".png"):
        slug = slug[:-4]
    if not slug:
        return None
    return f"{_STEAM_CDN}/abilities/{slug}.png"


def item_icon_url(icon_path: str | None) -> str | None:
    """Build Steam CDN item icon URL from dotabase icon path."""
    if not icon_path:
        return None
    # e.g. /panorama/images/items/blink_png.png
    slug = icon_path.replace("/panorama/images/items/", "").replace("_png.png", "")
    if slug.endswith(".png"):
        slug = slug[:-4]
    if not slug:
        return None
    return f"{_STEAM_CDN}/items/{slug}.png"


# ---------------------------------------------------------------------------
# Data loaders (cached)
# ---------------------------------------------------------------------------

@lru_cache(maxsize=1)
def load_heroes() -> list[HeroData]:
    session = dotabase_session()
    heroes = session.query(Hero).all()
    result = []
    for h in heroes:
        agi_base = h.attr_agility_base or 0
        base_a = h.base_armor if h.base_armor is not None else 0
        result.append(HeroData(
            id=h.id,
            name=h.name or "",
            localized_name=h.localized_name or "",
            real_name=h.real_name if h.real_name else None,
            hype=h.hype if h.hype else None,
            bio=h.bio if h.bio else None,
            attr_primary=h.attr_primary,
            is_melee=bool(h.is_melee),
            base_movement=h.base_movement,
            base_armor=h.base_armor,
            attack_rate=h.attack_rate if h.attack_rate else None,
            attr_str_gain=h.attr_strength_gain if h.attr_strength_gain else None,
            attr_agi_gain=h.attr_agility_gain if h.attr_agility_gain else None,
            attr_int_gain=h.attr_intelligence_gain if h.attr_intelligence_gain else None,
            image_url=hero_image_url(h.name or ""),
            armor_at_level1=round(base_a + agi_base / 6, 4),
            vision_night=h.vision_night if h.vision_night else None,
            turn_rate=h.turn_rate if h.turn_rate else None,
            attack_damage_min=h.attack_damage_min if h.attack_damage_min else None,
            attack_damage_max=h.attack_damage_max if h.attack_damage_max else None,
        ))
    return result


@lru_cache(maxsize=1)
def load_abilities() -> list[AbilityData]:
    session = dotabase_session()
    abilities = session.query(Ability).options(joinedload(Ability.hero)).all()
    result = []
    for a in abilities:
        if a.is_talent:
            continue
        name = a.localized_name
        if not name:
            continue
        if "_" in name:
            continue  # skip internal/hidden abilities (e.g., rubick_hidden3)
        hero_name = a.hero.localized_name if a.hero else None
        result.append(AbilityData(
            id=a.id,
            name=a.name or "",
            localized_name=name,
            hero_id=a.hero_id,
            hero_name=hero_name,
            damage_type=a.damage_type if a.damage_type else None,
            damage=a.damage if a.damage and a.damage != "0" else None,
            cooldown=a.cooldown if a.cooldown and a.cooldown != "0" else None,
            lore=a.lore if a.lore else None,
            scepter_upgrades=bool(a.scepter_upgrades),
            scepter_description=a.scepter_description if a.scepter_description else None,
            shard_upgrades=bool(a.shard_upgrades),
            shard_description=a.shard_description if a.shard_description else None,
            innate=bool(a.innate),
            icon_url=ability_icon_url(a.icon),
            mana_cost=a.mana_cost if a.mana_cost and a.mana_cost != "0" else None,
        ))
    return result


@lru_cache(maxsize=1)
def load_items() -> list[ItemData]:
    session = dotabase_session()
    items = session.query(Item).all()

    # First pass: filter unavailable items and extract base_level for suffix logic
    raw: list[tuple[Item, int | None]] = []
    for i in items:
        if not i.localized_name or "_" in i.localized_name:
            continue
        item_json = json.loads(i.json_data) if isinstance(i.json_data, str) else (i.json_data or {})
        # Skip items disabled in Valve's game files (removed/unavailable items like Trident,
        # Iron Talon, etc.). Neutral drops are exempt — they're obtained, not purchased.
        if i.neutral_tier is None and item_json.get("ItemPurchasable") in (0, "0"):
            continue
        base_level = item_json.get("ItemBaseLevel")
        raw.append((i, int(base_level) if base_level is not None else None))

    # Items whose localized_name appears at multiple base levels need a level suffix
    # (currently only Dagon 1-5 all share the name "Dagon")
    name_counts = Counter(i.localized_name for i, _ in raw)

    # Second pass: build ItemData with level suffix applied where needed
    result = []
    for i, base_level in raw:
        name = i.localized_name
        if name_counts[name] > 1 and base_level is not None:
            level_str = f" {base_level}"
            if not name.endswith(level_str):
                name = name + level_str
        result.append(ItemData(
            id=i.id,
            localized_name=name,
            cost=i.cost if i.cost and i.cost > 0 else None,
            lore=i.lore if i.lore else None,
            neutral_tier=i.neutral_tier,
            icon_url=item_icon_url(i.icon),
            is_neutral_enhancement=bool(getattr(i, "is_neutral_enhancement", False)),
            ability_special=i.ability_special if i.ability_special else None,
            active_cooldown=i.cooldown if i.cooldown and i.cooldown != "0" else None,
        ))
    return result


@lru_cache(maxsize=1)
def load_voicelines() -> list[VoicelineData]:
    """Load voicelines suitable for trivia (clean, reasonable length)."""
    session = dotabase_session()
    responses = (
        session.query(Response)
        .filter(Response.hero_id.isnot(None), Response.text_simple.isnot(None))
        .all()
    )
    result = []
    for r in responses:
        text = (r.text_simple or "").strip()
        # Filter: reasonable length, no hero name leak via criteria check
        if len(text) < 15 or len(text) > 120:
            continue
        # Skip generic/boring lines
        if text.lower() in {"", "hahaha", "ha ha ha"}:
            continue
        result.append(VoicelineData(hero_id=r.hero_id, text=text))
    return result


@lru_cache(maxsize=1)
def load_facets() -> list[FacetData]:
    session = dotabase_session()
    heroes = session.query(Hero).options(joinedload(Hero.facets)).all()
    result = []
    for h in heroes:
        if not h.facets:
            continue
        for f in h.facets:
            if not f.localized_name or not f.description:
                continue
            result.append(FacetData(
                id=f.id,
                localized_name=f.localized_name,
                hero_id=h.id,
                hero_name=h.localized_name,
                description=f.description,
            ))
    return result


# ---------------------------------------------------------------------------
# Lookup helpers
# ---------------------------------------------------------------------------

_hero_by_id: dict[int, HeroData] = {}


def get_hero_by_id(hero_id: int) -> HeroData | None:
    if not _hero_by_id:
        for h in load_heroes():
            _hero_by_id[h.id] = h
    return _hero_by_id.get(hero_id)


def redact_hero_name(text: str, hero_name: str) -> str:
    """Remove hero name references from text for lore/bio questions."""
    if not text or not hero_name:
        return text or ""
    # Redact full name and individual words (for multi-word names)
    result = re.sub(re.escape(hero_name), "???", text, flags=re.IGNORECASE)
    for word in hero_name.split():
        if len(word) > 2:  # Don't redact tiny words like "of"
            result = re.sub(r"\b" + re.escape(word) + r"\b", "???", result, flags=re.IGNORECASE)
    return result

File: rust/crates/cama-app/src/trivia_data.rs
//! Dotabase-independent trivia catalog loading and normalization.
//!
//! The immutable catalog reader is represented by [`TriviaDataSource`]; this module owns only the
//! filtering, derived-stat, URL, caching, lookup, and redaction policy applied
//! after a source has returned owned rows. No application database is opened.

use std::sync::OnceLock;

use thiserror::Error;

use crate::trivia_questions::{
    AbilityData as QuestionAbilityData, HeroData as QuestionHeroData, ItemData as QuestionItemData,
    ItemSpecialValue, TriviaCatalog, VoicelineData as QuestionVoicelineData,
};

const STEAM_CDN: &str = "https://cdn.cloudflare.steamstatic.com/apps/dota2/images/dota_react";

#[derive(Clone, Debug, Default, PartialEq)]
pub struct RawHero {
    pub id: i64,
    pub name: Option<String>,
    pub localized_name: Option<String>,
    pub real_name: Option<String>,
    pub hype: Option<String>,
    pub bio: Option<String>,
    pub primary_attribute: Option<String>,
    pub is_melee: bool,
    pub base_movement: Option<i64>,
    pub base_armor: Option<f64>,
    pub attack_rate: Option<f64>,
    pub strength_base: Option<f64>,
    pub agility_base: Option<f64>,
    pub intelligence_base: Option<f64>,
    pub strength_gain: Option<f64>,
    pub agility_gain: Option<f64>,
    pub intelligence_gain: Option<f64>,
    pub night_vision: Option<i64>,
    pub turn_rate: Option<f64>,
    pub attack_damage_min: Option<i64>,
    pub attack_damage_max: Option<i64>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct RawAbility {
    pub id: i64,
    pub name: Option<String>,
    pub localized_name: Option<String>,
    pub is_talent: bool,
    pub hero_id: Option<i64>,
    pub hero_name: Option<String>,
    pub damage_type: Option<String>,
    pub damage: Option<String>,
    pub cooldown: Option<String>,
    pub lore: Option<String>,
    pub scepter_upgrades: bool,
    pub scepter_description: Option<String>,
    pub shard_upgrades: bool,
    pub shard_description: Option<String>,
    pub innate: bool,
    pub icon_path: Option<String>,
    pub mana_cost: Option<String>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct RawItem {
    pub id: i64,
    pub localized_name: Option<String>,
    pub cost: Option<i64>,
    pub lore: Option<String>,
    pub neutral_tier: Option<i64>,
    pub icon_path: Option<String>,
    pub is_neutral_enhancement: bool,
    pub special_values: Vec<ItemSpecialValue>,
    pub active_cooldown: Option<String>,
    /// Parsed `ItemPurchasable`; neutral drops are exempt when false.
    pub purchasable: Option<bool>,
    /// Parsed `ItemBaseLevel` from Valve's item JSON.
    pub base_level: Option<i64>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct RawVoiceline {
    pub hero_id: Option<i64>,
    pub text: Option<String>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct RawFacet {
    pub id: i64,
    pub localized_name: Option<String>,
    pub hero_id: i64,
    pub hero_name: Option<String>,
    pub description: Option<String>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct HeroData {
    pub id: i64,
    pub name: String,
    pub localized_name: String,
    pub real_name: Option<String>,
    pub hype: Option<String>,
    pub bio: Option<String>,
    pub primary_attribute: Option<String>,
    pub is_melee: bool,
    pub base_movement: Option<i64>,
    pub base_armor: Option<f64>,
    pub attack_rate: Option<f64>,
    pub strength_gain: Option<f64>,
    pub agility_gain: Option<f64>,
    pub intelligence_gain: Option<f64>,
    pub image_url: Option<String>,
    pub armor_at_level_one: f64,
    pub night_vision: Option<i64>,
    pub turn_rate: Option<f64>,
    pub attack_damage_min: Option<i64>,
    pub attack_damage_max: Option<i64>,
    pub damage_min_at_level_one: Option<i64>,
    pub damage_max_at_level_one: Option<i64>,
}

impl From<&HeroData> for QuestionHeroData {
    fn from(value: &HeroData) -> Self {
        Self {
            id: value.id,
            localized_name: value.localized_name.clone(),
            real_name: value.real_name.clone(),
            hype: value.hype.clone(),
            bio: value.bio.clone(),
            primary_attribute: value.primary_attribute.clone(),
            is_melee: value.is_melee,
            base_movement: value.base_movement,
            attack_rate: value.attack_rate,
            strength_gain: value.strength_gain,
            agility_gain: value.agility_gain,
            intelligence_gain: value.intelligence_gain,
            image_url: value.image_url.clone(),
            armor_at_level_one: Some(value.armor_at_level_one),
            night_vision: value.night_vision,
            turn_rate: value.turn_rate,
            damage_min_at_level_one: value.damage_min_at_level_one,
            damage_max_at_level_one: value.damage_max_at_level_one,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AbilityData {
    pub id: i64,
    pub name: String,
    pub localized_name: String,
    pub hero_id: Option<i64>,
    pub hero_name: Option<String>,
    pub damage_type: Option<String>,
    pub damage: Option<String>,
    pub cooldown: Option<String>,
    pub lore: Option<String>,
    pub scepter_upgrades: bool,
    pub scepter_description: Option<String>,
    pub shard_upgrades: bool,
    pub shard_description: Option<String>,
    pub innate: bool,
    pub icon_url: Option<String>,
    pub mana_cost: Option<String>,
}

impl From<&AbilityData> for QuestionAbilityData {
    fn from(value: &AbilityData) -> Self {
        Self {
            localized_name: value.localized_name.clone(),
            hero_id: value.hero_id,
            hero_name: value.hero_name.clone(),
            damage_type: value.damage_type.clone(),
            damage: value.damage.clone(),
            cooldown: value.cooldown.clone(),
            lore: value.lore.clone(),
            scepter_upgrades: value.scepter_upgrades,
            scepter_description: value.scepter_description.clone(),
            shard_upgrades: value.shard_upgrades,
            shard_description: value.shard_description.clone(),
            innate: value.innate,
            icon_url: value.icon_url.clone(),
            mana_cost: value.mana_cost.clone(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ItemData {
    pub id: i64,
    pub localized_name: String,
    pub cost: Option<i64>,
    pub lore: Option<String>,
    pub neutral_tier: Option<i64>,
    pub icon_url: Option<String>,
    pub is_neutral_enhancement: bool,
    pub special_values: Vec<ItemSpecialValue>,
    pub active_cooldown: Option<String>,
    pub base_level: Option<i64>,
    pub is_leveled: bool,
}

impl From<&ItemData> for QuestionItemData {
    fn from(value: &ItemData) -> Self {
        Self {
            localized_name: value.localized_name.clone(),
            cost: value.cost,
            lore: value.lore.clone(),
            neutral_tier: value.neutral_tier,
            icon_url: value.icon_url.clone(),
            special_values: value.special_values.clone(),
            active_cooldown: value.active_cooldown.clone(),
            is_leveled: value.is_leveled,
        }
    }
}

pub type VoicelineData = QuestionVoicelineData;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FacetData {
    pub id: i64,
    pub localized_name: String,
    pub hero_id: i64,
    pub hero_name: Option<String>,
    pub description: String,
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
pub enum TriviaDataError {
    #[error("trivia data source failed: {0}")]
    Source(String),
}

/// Asset-reader boundary. Implementations acquire and release any database
/// connection inside each method and return detached, owned rows.
pub trait TriviaDataSource {
    fn heroes(&self) -> Result<Vec<RawHero>, TriviaDataError>;
    fn abilities(&self) -> Result<Vec<RawAbility>, TriviaDataError>;
    fn items(&self) -> Result<Vec<RawItem>, TriviaDataError>;
    fn voicelines(&self) -> Result<Vec<RawVoiceline>, TriviaDataError>;
    fn facets(&self) -> Result<Vec<RawFacet>, TriviaDataError>;
}

/// Process-lifetime cache equivalent to Python's one-entry loader caches.
#[derive(Debug)]
pub struct TriviaDataRegistry<S> {
    source: S,
    heroes: OnceLock<Vec<HeroData>>,
    abilities: OnceLock<Vec<AbilityData>>,
    items: OnceLock<Vec<ItemData>>,
    voicelines: OnceLock<Vec<VoicelineData>>,
    facets: OnceLock<Vec<FacetData>>,
}

impl<S> TriviaDataRegistry<S> {
    #[must_use]
    pub const fn new(source: S) -> Self {
        Self {
            source,
            heroes: OnceLock::new(),
            abilities: OnceLock::new(),
            items: OnceLock::new(),
            voicelines: OnceLock::new(),
            facets: OnceLock::new(),
        }
    }

    #[must_use]
    pub const fn source(&self) -> &S {
        &self.source
    }

    /// Mirrors Python's `cache_clear` hooks for tests and asset refreshes.
    pub fn clear_caches(&mut self) {
        self.heroes = OnceLock::new();
        self.abilities = OnceLock::new();
        self.items = OnceLock::new();
        self.voicelines = OnceLock::new();
        self.facets = OnceLock::new();
    }
}

impl<S: TriviaDataSource> TriviaDataRegistry<S> {
    pub fn load_heroes(&self) -> Result<&[HeroData], TriviaDataError> {
        cached(&self.heroes, || self.source.heroes().map(normalize_heroes))
    }

    pub fn load_abilities(&self) -> Result<&[AbilityData], TriviaDataError> {
        cached(&self.abilities, || {
            self.source.abilities().map(normalize_abilities)
        })
    }

    pub fn load_items(&self) -> Result<&[ItemData], TriviaDataError> {
        cached(&self.items, || self.source.items().map(normalize_items))
    }

    pub fn load_voicelines(&self) -> Result<&[VoicelineData], TriviaDataError> {
        cached(&self.voicelines, || {
            self.source.voicelines().map(normalize_voicelines)
        })
    }

    pub fn load_facets(&self) -> Result<&[FacetData], TriviaDataError> {
        cached(&self.facets, || self.source.facets().map(normalize_facets))
    }

    pub fn get_ability_icon_url_by_name(
        &self,
        localized_name: &str,
    ) -> Result<Option<&str>, TriviaDataError> {
        let target = localized_name.trim().to_lowercase();
        if target.is_empty() {
            return Ok(None);
        }
        Ok(self
            .load_abilities()?
            .iter()
            .find(|ability| ability.localized_name.to_lowercase() == target)
            .and_then(|ability| ability.icon_url.as_deref()))
    }

    pub fn get_hero_by_id(&self, hero_id: i64) -> Result<Option<&HeroData>, TriviaDataError> {
        Ok(self.load_heroes()?.iter().find(|hero| hero.id == hero_id))
    }

    /// Reuse the existing question engine's catalog types without duplicating
    /// its generator registry.
    pub fn question_catalog(&self) -> Result<TriviaCatalog, TriviaDataError> {
        Ok(TriviaCatalog {
            heroes: self.load_heroes()?.iter().map(Into::into).collect(),
            abilities: self.load_abilities()?.iter().map(Into::into).collect(),
            items: self.load_items()?.iter().map(Into::into).collect(),
            voicelines: self.load_voicelines()?.to_vec(),
        })
    }
}

fn cached<T>(
    cell: &OnceLock<Vec<T>>,
    load: impl FnOnce() -> Result<Vec<T>, TriviaDataError>,
) -> Result<&[T], TriviaDataError> {
    if let Some(values) = cell.get() {
        return Ok(values);
    }
    let loaded = load()?;
    let _ = cell.set(loaded);
    Ok(cell
        .get()
        .expect("either this caller or a concurrent caller initialized the cache"))
}

#[must_use]
pub fn hero_image_url(hero_name: &str) -> Option<String> {
    let slug = hero_name.replace("npc_dota_hero_", "");
    (!slug.is_empty()).then(|| format!("{STEAM_CDN}/heroes/{slug}.png"))
}

#[must_use]
pub fn ability_icon_url(icon_path: Option<&str>) -> Option<String> {
    icon_url(icon_path, "/panorama/images/spellicons/", "abilities")
}

#[must_use]
pub fn item_icon_url(icon_path: Option<&str>) -> Option<String> {
    icon_url(icon_path, "/panorama/images/items/", "items")
}

fn icon_url(icon_path: Option<&str>, prefix: &str, category: &str) -> Option<String> {
    let path = icon_path.filter(|path| !path.is_empty())?;
    let mut slug = path.replace(prefix, "").replace("_png.png", "");
    if slug.ends_with(".png") {
        slug.truncate(slug.len() - 4);
    }
    (!slug.is_empty()).then(|| format!("{STEAM_CDN}/{category}/{slug}.png"))
}

#[must_use]
pub fn redact_hero_name(text: Option<&str>, hero_name: &str) -> String {
    let Some(text) = text.filter(|text| !text.is_empty()) else {
        return String::new();
    };
    if hero_name.is_empty() {
        return text.to_owned();
    }
    crate::trivia_questions::redact_hero_name(text, hero_name)
}

fn normalize_heroes(rows: Vec<RawHero>) -> Vec<HeroData> {
    rows.into_iter()
        .map(|hero| {
            let strength = hero.strength_base.unwrap_or(0.0);
            let agility = hero.agility_base.unwrap_or(0.0);
            let intelligence = hero.intelligence_base.unwrap_or(0.0);
            let bonus = level_one_attribute_bonus(
                hero.primary_attribute.as_deref(),
                strength,
                agility,
                intelligence,
            );
            let damage_min_at_level_one = hero
                .attack_damage_min
                .filter(|damage| *damage != 0)
                .map(|damage| (damage as f64 + bonus).round_ties_even() as i64);
            let damage_max_at_level_one = hero
                .attack_damage_max
                .filter(|damage| *damage != 0)
                .map(|damage| (damage as f64 + bonus).round_ties_even() as i64);
            let name = hero.name.unwrap_or_default();
            HeroData {
                id: hero.id,
                image_url: hero_image_url(&name),
                name,
                localized_name: hero.localized_name.unwrap_or_default(),
                real_name: nonempty(hero.real_name),
                hype: nonempty(hero.hype),
                bio: nonempty(hero.bio),
                primary_attribute: hero.primary_attribute,
                is_melee: hero.is_melee,
                base_movement: hero.base_movement,
                base_armor: hero.base_armor,
                attack_rate: nonzero_f64(hero.attack_rate),
                strength_gain: nonzero_f64(hero.strength_gain),
                agility_gain: nonzero_f64(hero.agility_gain),
                intelligence_gain: nonzero_f64(hero.intelligence_gain),
                armor_at_level_one: round_four(hero.base_armor.unwrap_or(0.0) + agility / 6.0),
                night_vision: nonzero_i64(hero.night_vision),
                turn_rate: nonzero_f64(hero.turn_rate),
                attack_damage_min: nonzero_i64(hero.attack_damage_min),
                attack_damage_max: nonzero_i64(hero.attack_damage_max),
                damage_min_at_level_one,
                damage_max_at_level_one,
            }
        })
        .collect()
}

fn normalize_abilities(rows: Vec<RawAbility>) -> Vec<AbilityData> {
    rows.into_iter()
        .filter_map(|ability| {
            let localized_name = ability.localized_name?;
            if ability.is_talent || localized_name.is_empty() || localized_name.contains('_') {
                return None;
            }
            Some(AbilityData {
                id: ability.id,
                name: ability.name.unwrap_or_default(),
                localized_name,
                hero_id: ability.hero_id,
                hero_name: nonempty(ability.hero_name),
                damage_type: nonempty(ability.damage_type),
                damage: nonzero_text(ability.damage),
                cooldown: nonzero_text(ability.cooldown),
                lore: nonempty(ability.lore),
                scepter_upgrades: ability.scepter_upgrades,
                scepter_description: nonempty(ability.scepter_description),
                shard_upgrades: ability.shard_upgrades,
                shard_description: nonempty(ability.shard_description),
                innate: ability.innate,
                icon_url: ability_icon_url(ability.icon_path.as_deref()),
                mana_cost: nonzero_text(ability.mana_cost),
            })
        })
        .collect()
}

fn normalize_items(rows: Vec<RawItem>) -> Vec<ItemData> {
    let filtered = rows
        .into_iter()
        .filter_map(|item| {
            let name = item.localized_name.as_deref()?;
            if name.is_empty() || name.contains('_') {
                return None;
            }
            if item.neutral_tier.is_none() && item.purchasable == Some(false) {
                return None;
            }
            Some(item)
        })
        .collect::<Vec<_>>();
    let mut name_counts = std::collections::BTreeMap::new();
    for item in &filtered {
        if let Some(name) = item.localized_name.as_ref() {
            *name_counts.entry(name.clone()).or_insert(0_usize) += 1;
        }
    }
    filtered
        .into_iter()
        .map(|item| {
            let mut localized_name = item.localized_name.unwrap_or_default();
            let is_leveled = name_counts.get(&localized_name).copied().unwrap_or(0) > 1
                && item.base_level.is_some();
            if let Some(level) = item.base_level.filter(|_| is_leveled) {
                let suffix = format!(" {level}");
                if !localized_name.ends_with(&suffix) {
                    localized_name.push_str(&suffix);
                }
            }
            ItemData {
                id: item.id,
                localized_name,
                cost: item.cost.filter(|cost| *cost > 0),
                lore: nonempty(item.lore),
                neutral_tier: item.neutral_tier,
                icon_url: item_icon_url(item.icon_path.as_deref()),
                is_neutral_enhancement: item.is_neutral_enhancement,
                special_values: item.special_values,
                active_cooldown: nonzero_text(item.active_cooldown),
                base_level: item.base_level,
                is_leveled,
            }
        })
        .collect()
}

fn normalize_voicelines(rows: Vec<RawVoiceline>) -> Vec<VoicelineData> {
    rows.into_iter()
        .filter_map(|voiceline| {
            let hero_id = voiceline.hero_id?;
            let text = voiceline.text.unwrap_or_default().trim().to_owned();
            let length = text.chars().count();
            if !(15..=120).contains(&length)
                || matches!(text.to_lowercase().as_str(), "" | "hahaha" | "ha ha ha")
            {
                return None;
            }
            Some(VoicelineData { hero_id, text })
        })
        .collect()
}

fn normalize_facets(rows: Vec<RawFacet>) -> Vec<FacetData> {
    rows.into_iter()
        .filter_map(|facet| {
            let localized_name = facet.localized_name.filter(|value| !value.is_empty())?;
            let description = facet.description.filter(|value| !value.is_empty())?;
            Some(FacetData {
                id: facet.id,
                localized_name,
                hero_id: facet.hero_id,
                hero_name: nonempty(facet.hero_name),
                description,
            })
        })
        .collect()
}

fn level_one_attribute_bonus(
    primary: Option<&str>,
    strength: f64,
    agility: f64,
    intelligence: f64,
) -> f64 {
    match primary {
        Some("strength") => strength,
        Some("agility") => agility,
        Some("intelligence") => intelligence,
        Some("universal") => 0.7 * (strength + agility + intelligence),
        _ => 0.0,
    }
}

fn round_four(value: f64) -> f64 {
    (value * 10_000.0).round_ties_even() / 10_000.0
}

fn nonempty(value: Option<String>) -> Option<String> {
    value.filter(|value| !value.is_empty())
}

fn nonzero_text(value: Option<String>) -> Option<String> {
    value.filter(|value| !value.is_empty() && value != "0")
}

fn nonzero_i64(value: Option<i64>) -> Option<i64> {
    value.filter(|value| *value != 0)
}

fn nonzero_f64(value: Option<f64>) -> Option<f64> {
    value.filter(|value| *value != 0.0)
}

#[cfg(test)]
mod tests;
